Add CD-grid stress arrays

phil-blain opened the pull request:
to be discussed... I tried grepping 'stressp_1' and added the new arrays where I thought they were needed... I'll clean up the commit history tomorrow. Probably missing some stuff

eclare108213 reviewed (commented):
Looks good so far.  Thanks for implementing these arrays in the restart files!

dabail10 commented on cicecore/cicedynB/analysis/ice_history.F90:
```diff
@@ -4287,17 +4288,29 @@ subroutine accum_hist (dt)
+                                      sig1      (:,:,iblk), &
+                                      sig2      (:,:,iblk), &
+                                      sigP      (:,:,iblk))
+            case('CD')
```
Why is there an extra case('CD') here?

phil-blain replied:
fixed.

dabail10 commented on cicecore/cicedynB/general/ice_flux.F90:
```diff
@@ -135,6 +135,9 @@ module ice_flux
          stressp_1, stressp_2, stressp_3, stressp_4 , & ! sigma11+sigma22
          stressm_1, stressm_2, stressm_3, stressm_4 , & ! sigma11-sigma22
          stress12_1,stress12_2,stress12_3,stress12_4    ! sigma12
```
This line needs a comma and ampersand.

phil-blain replied:
also fixed

dabail10 reviewed (changes requested):
This is looking really good. There might be some syntax errors beyond the one that github actions failed on.

phil-blain commented:
Yeah I pushed that without even compiling (!). I compiles now, I'll update the PR soon.

phil-blain committed:
ice_flux: add stress arrays at CD-grid locations (T,U)

phil-blain committed:
ice_dyn_shared: initialize CD-grid stress arrays

phil-blain committed:
infrastructure: add CD-grid stress arrays to restarts

phil-blain committed:
ice_dyn_shared: generalize 'principal_stress' arguments names

In a subsequent commit we will call 'principal_stress' with the CD-grid
stress arrays 'stress{p,m,12}T' to compute the principal stresses at the
tracer point when using the CD grid.

In that light, remove the '_1' suffix from the stress arguments since
they won't always be located at the NE corner anymore.

phil-blain committed:
ice_history: compute 'sig[12P]' at T-point for CD-grid

For the CD-grid, compute the principal stresses and the ice pressure at
the T point by passing the appropriate arrays to 'principal_stress'.

These three history variables are computed using the NE-corner values on
the B-grid, but this is not reflected in the description of the fields
in the history output.

Add the location at which the stresses are computed to the comment
argument in the call to 'define_hist_field', for both the B and CD grid.

phil-blain committed:
ice_dyn_evp: pass 'stress{p,m,12}[TU]' to dyn_prep2

For the CD-grid, the 'stress{p,m,12}_[1-4]' arrays are not used. Pass
the CD-grid location stress arrays 'stress{p,m,12}[TU]' to 'dyn_prep2',
which zero-initializes them anywhere icetmask is zero.

Also, add a TODO in the `if (grid_type) == 'tripole'` block, since it is
not yet clear how the halo updates should be done for the new stress
arrays on the tripole grid.

phil-blain commented:
I cleaned up the compile failures, and cleaned up the commit history.

There are ~two~ three TODOs that maybe we will want to discuss..

phil-blain commented on cicecore/cicedynB/infrastructure/ice_restart_driver.F90:
```diff
@@ -465,6 +495,7 @@ subroutine restartfile (ice_ic)
             stress12_4(i,j,iblk) = c0
          enddo
          enddo
+         ! TODO: CD-grid ?
```
These initialization date from https://github.com/CICE-Consortium/CICE-svn-trunk/commit/788626f9
